File: src/version c++/flowshop.cpp
//
// Created by Alexandre on 0026 26 mars 2021.
//

/***************************************************************************
 *   Copyright (C) 2012 by Jérémie Dubois-Lacoste   *
 *   jeremie.dl@gmail.com   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <iostream>
#include <cstdlib>
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <algorithm>
#include <bits/stdc++.h>
#include "pfspinstance.h"
#include "flowshop.h"

using namespace std;

void printVector(vector<int> v, string message) {
    cout << message;
    for (int i=0; i<v.size(); i++){
        cout << v[i] << " ";
    }
    cout << endl;
}

void printSol(Solution s) {
    printVector(s.sol, "jobs : ");
    cout << "wct : " << s.wct << endl;
}

/***************************************************************************/

Solution generateRndSol(PfspInstance instance) {
    /*
    Fill the solution with numbers between 0 and nb_jobs, shuffled
    */
    std::vector<int> sol(instance.getNbJob());
    for (int i=0; i<instance.getNbJob(); i++) {
        sol[i] = i;
    }
    random_shuffle(sol.begin(), sol.end());
    Solution random_sol = {sol, instance.computeWCT(sol)};
    return random_sol;
}


vector<int> getInitSRZsol(PfspInstance instance) {
    /*
    Get the set of jobs ordered (ascending) according to their
    weighted sum of processing times
    */
    multimap<double, int > mymap; // sorted map in ascending order of the keys

    for (int i = 0; i < instance.getNbJob(); i++) {
        mymap.insert(make_pair(instance.weightedSumSingleJob(i), i));
    }

    vector<int> init_sol(0);
    // begin() returns to the first value of multimap.
    multimap<double,int> :: iterator it;
    for (it = mymap.begin() ; it != mymap.end() ; it++) {
        init_sol.push_back((*it).second);
//        cout << "(" << (*it).first << ", " << (*it).second << ")" << endl;
    }
    return init_sol;
}

Solution getBestSubset(int size, int job, vector<int> start, PfspInstance instance) {
    vector<int> subset = start;

    Solution best_subset = {vector<int>(0), static_cast<long>(10000000000)};
    for (int i = 0; i < size; i++){
        auto it = subset.begin();
        it = subset.insert ( it+i, job ); // insert job to position i
        int temp_wct = instance.computeWCT(subset);
//        printVector(subset, "subset : ");
//        cout << "=> " << temp_wct << endl;

        if (temp_wct <= best_subset.wct) {
            best_subset.sol = subset;
            best_subset.wct = temp_wct;
        }
        if (i < subset.size()) {
            subset.erase(subset.begin()+i); // remove the ith element of the list
        }
    }
    return best_subset;
}

Solution simplifiedRZheuristic(PfspInstance instance) {
    /*
    Start by ordering the jobs in ascending order of their weighted sum of processing times
	Then, construct the solution by inserting one job at a time in the position that minimize the WCT
    */
    vector<int> starting_sequence = getInitSRZsol(instance);
//    printVector(starting_sequence, "starting sequence : ");
    Solution best_sol = {vector<int>(0), static_cast<long>(10000000000)};
    for (int j = 0; j < instance.getNbJob(); j++) {
        Solution best_subset = getBestSubset(j+1, starting_sequence[j], best_sol.sol, instance);
//        printSol(best_subset);
        best_sol.sol = best_subset.sol;
        best_sol.wct = best_subset.wct;
    }
    return best_sol;

}

/***************************************************************************/

Solution getBestTransposeNeighbour(Solution sol, PfspInstance instance, string pivoting_rule) {
    /*
    Two permutations ø, ø' are transpose neighbours if, and only if,
	one can be obtained from the other by swapping the positions of two adjacent jobs
	ex : [A,B,C,D,E,F]
	=>   [A,C,B,D,E,F]
    */
    Solution best_sol = {vector<int>(0), sol.wct};
    vector<int> temp_sol;

    for (int i = 0; i < sol.sol.size()-1; i++) {
        temp_sol = sol.sol;
        iter_swap(temp_sol.begin() + i, temp_sol.begin() + i+1);
//        printVector(temp_sol, "");
        int wct = instance.computeWCT(temp_sol);
        if (wct < best_sol.wct) {
            best_sol.sol = temp_sol;
            best_sol.wct = wct;
            if (pivoting_rule == "FI") {
                return best_sol;
            }
        }
    }
    return best_sol;

}

Solution getBestExchangeNeighbour(Solution sol, PfspInstance instance, string pivoting_rule) {
    /*
    Two permutations ø, ø' are 2-exchange neighbours if, and only if,
    one can be obtained from the other by exchanging two jobs at arbitrary positions
    ex : [A,B,C,D,E,F]
    =>   [A,E,C,D,B,F]
    */
    Solution best_sol = {vector<int>(0), sol.wct};
    vector<int> temp_sol;

    for (int i = 0; i < sol.sol.size(); i++) {
        for (int j = 0; j < sol.sol.size(); j++) {
            if (i != j) {
                temp_sol = sol.sol;
                iter_swap(temp_sol.begin() + i, temp_sol.begin() + j);
//                printVector(temp_sol, "");
                int wct = instance.computeWCT(temp_sol);
                if (wct < best_sol.wct) {
                    best_sol.sol = temp_sol;
                    best_sol.wct = wct;
                    if (pivoting_rule == "FI") {
                        return best_sol;
                    }
                }
            }
        }
    }
    return best_sol;
}

Solution getBestInsertionNeighbour(Solution sol, PfspInstance instance, string pivoting_rule) {
    /*
    Two permutations ø, ø' are insertion neighbours if, and only if,
    one can be obtained from the other by removing a job from one position
                                                                  and inserting it at another position
    ex : [A,B,C,D,E,F]
                 =>   [A,C,D,B,E,F]

    QUESTION : l'insertion peuut se faire en fin de liste aussi ?
    */
    Solution best_sol = {vector<int>(0), sol.wct};
    vector<int> temp_sol;

    for (int i = 0; i < sol.sol.size(); i++) {
        for (int j = 0; j < sol.sol.size() + 1; j++) {
            if (i != j and j != i+1) {
                temp_sol = sol.sol;
                auto it = temp_sol.begin();
                it = temp_sol.insert ( it+j, sol.sol[i] ); // insert job to position j
                if (j < i) {
                    temp_sol.erase(temp_sol.begin()+i+1); // remove the (i+1)th element of the list
                } else {
                    temp_sol.erase(temp_sol.begin()+i); // remove the ith element of the list
                }
//                printVector(temp_sol, "");
                int wct = instance.computeWCT(temp_sol);
                if (wct < best_sol.wct) {
                    best_sol.sol = temp_sol;
                    best_sol.wct = wct;
                    if (pivoting_rule == "FI") {
                        return best_sol;
                    }
                }
            }
        }
    }
    return best_sol;
}

/***************************************************************************/


// ITERATIVE IMPROVEMENT
bool isLocalOptimal(Solution sol) {
    /*
    Check if the solution is a local optimum. By definition, a local optimum (here minimum) is :
    search position without improving neighbours w.r.t. given evaluation function g and neighbourhood
    N. i.e. position s in S such that g(s) <= g(s') for all s' of N(s).
    => same as checking if the neighbour found is empty. (if empty : sol is a local optimum)
    */
    return (sol.sol.empty()); // if true => local optimum
}

Solution generateInitialSolution(string mode, PfspInstance instance) {
    if (mode == "R") {
        return generateRndSol(instance);
    } else if (mode == "SRZ") {
        return simplifiedRZheuristic(instance);
    }
}

Solution chooseNeighbour(Solution sol, PfspInstance instance, string neighbour_type, string pivoting_rule) {
    /*
    Choose one neighbour (the best one) among all the neighbours of a given solution.
    */
    if (neighbour_type == "T") {
        return getBestTransposeNeighbour(sol, instance, pivoting_rule);
    } else if (neighbour_type == "E") {
        return getBestExchangeNeighbour(sol, instance, pivoting_rule);
    } else if (neighbour_type == "I"){
        return getBestInsertionNeighbour(sol, instance, pivoting_rule);
    }
}

Solution iterativeImprovement(string pivoting_rule, string neigh_rule, PfspInstance instance, Solution init_sol) {
    /*
    Heuristic algorithm to find the optimal solution to the PSFP
    args = [pivotingrule, neighbourhood]
    */
    Solution best_sol = init_sol;
    Solution neighbour = chooseNeighbour(best_sol, instance, neigh_rule, pivoting_rule);
    int it = 0;
    while (!isLocalOptimal(neighbour)) {
        best_sol = neighbour;
        neighbour = chooseNeighbour(best_sol, instance, neigh_rule, pivoting_rule);
        it++;
    }
    return best_sol;
}

// VND : variable neighbourhood descent
Solution variableNeighbourhoodDescent(vector<string> neighbourhood_modes, PfspInstance instance, Solution init_sol) {
    /*
    VND variable neighbourhood descent
    use different neighbourhood mode throughout the algorithm
    args = [name.py, initsol, pivotingrule, neighbourhood]
    */
    int i = 0;
    int it = 0;
    bool stop = false;
    Solution best_sol = init_sol;
    while (not stop) { // "do while"
//        cout << "neighbourhood mode : " << neighbourhood_modes[i] << endl;
        Solution neighbour = chooseNeighbour(best_sol, instance, neighbourhood_modes[i], "FI");
        if (isLocalOptimal(neighbour)) { // if no existing improving solution
            i++;
        } else {
            best_sol = neighbour;
//            cout << "it " << it << " | ";
//            printSol(best_sol);
            i = 0;
            it++;
        }
        stop = i > neighbourhood_modes.size() - 1;
    }
    return best_sol;
}

vector<vector<int>> readBestSolFromFile(char * filename) {
    // TODO : must be working on ubuntu and windows with relative path

    bool everythingOK = true;
    double wct;
    string str;
    ifstream fileIn;

    vector<vector<int>> best_solutions = {vector<int>(), vector<int>()};

    fileIn.open(filename);

    if ( fileIn.is_open() ) {
        fileIn >> str; fileIn >> str; fileIn >> str; // "Problem , BS" => skip first line
        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 30; ++j) {
                fileIn >> str; // The instance name, not important !
                fileIn >> str; // "," not important
                fileIn >> wct; // best solution wct
                cout << wct << endl;
                best_solutions[i].push_back( (int) wct);
            }
        }

        cout << "All is read from file." << std::endl;
        fileIn.close();
    }
    else {
        cout    << "ERROR. file:pfspInstance.cpp, method:readBestSolFromFile, "
                << "error while opening file " << filename << endl;
        everythingOK = false;
    }

    return best_solutions;
}

File: src/flowshop.py
import pfsinstance
import random
import sys 
import time 
from operator import itemgetter
import copy

def generateRndPosition(mini, maxi):
	""" 
	Return a random number in the range [mini, maxi]
	"""
	return random.randint(mini, maxi)

def randomPermutation(nb_jobs):
	"""
	Fill the solution with numbers between 0 and nb_jobs, shuffled
	""" 
	# TODO : fix the generate random solution function (to have an uniform random generation)
	already_taken = [False for _ in range(nb_jobs)]
	choosen_nb = [0 for _ in range(nb_jobs)]
	sol = [-1 for _ in range(nb_jobs)]

	nbj = -1
	for i in range(nb_jobs-1, -1, -1):
		rnd = generateRndPosition(0, i)
		# print("rnd: ", rnd)
		nb_false = 0

		# find the rndth cell with value = false 
		j = 0
		while (nb_false < rnd):
			if (not already_taken[j]):
				nb_false += 1
			j+=1
		j-= 1
		# print(already_taken, j)

		sol[j] = i 

		nbj += 1 
		choosen_nb[nbj] = j

		already_taken[j] = True 

	return sol

def generateRndSol(instance):
	"""
	Fill the solution with numbers between 0 and nb_jobs, shuffled
	""" 
	sol = [i for i in range(instance.getNbJob())]
	random.shuffle(sol)
	return {"jobs": sol, "wct": instance.computeWCT3(sol)}

def getBestSubset(size, job, start, instance):
	# print("subset size : ", size)
	subset = start
	best_subset = []
	best_wct = 10000000000
	for i in range(size):
		subset.insert(i, job)

		temp_instance = pfsinstance.PfspInstance(len(subset), instance.getNbMac())
		temp_instance.setProcessingTimes(instance.getProcessingTimes())
		temp_instance.setWeights(instance.getWeights())
		wct = temp_instance.computeWCT3(subset)

		# print(">", subset, wct)
		if (wct <= best_wct):
			best_subset = copy.copy(subset) 
			best_wct = wct

		# print("subset before reset : ", subset)
		#reset the subset
		if (i < len(subset)):
			subset.pop(i)
		# subset = start
		# print("subset after reset : ", subset)

	return best_subset, best_wct



def simplifiedRZheuristic(instance):
	"""
	Start by ordering the jobs in ascending order of their weighted sum of processing times
	Then, construct the solution by inserting one job at a time in the position that minimize the WCT
	"""
	weighted_sums = {i: instance.weightedSumSingleJob(i) for i in range(instance.getNbJob())}
	starting_sequence= sorted(weighted_sums.items(), key=lambda x:x[1])
	
	# print("weighted sums: ", weighted_sums, len(weighted_sums))
	# print("starting sequence : ", starting_sequence, len(starting_sequence))

	best_sol = {"jobs":[], "wct": 100000000000} # step 1
	for j in range(len(starting_sequence)): # (job, T)
		# print("Step {0} = {1} with J{2}".format(j, best_sol, starting_sequence[j][0]))

		bests = getBestSubset(j+1, starting_sequence[j][0], best_sol["jobs"], instance)
		best_sol["jobs"] = bests[0]
		best_sol["wct"] = bests[1]
		# print("best : ", best_sol)


	return best_sol

def generateInitialSolution(mode, instance):
	if (mode == "R"):
		return generateRndSol(instance)
	elif (mode == "SRZ"):
		return simplifiedRZheuristic(instance)

def swapValues(temp_sol, i, j):
	# swap elements at position i and i + 1
	temp_val = temp_sol[i]
	temp_sol[i] = temp_sol[j]
	temp_sol[j] = temp_val 

def getBestTransposeNeighbour(sol, instance, pivoting_rule):
	"""
	Two permutations ø, ø' are transpose neighbours if, and only if, 
	one can be obtained from the other by swapping the positions of two adjacent jobs
	ex : [A,B,C,D,E,F]
	=>   [A,C,B,D,E,F]
	"""
	best_sol = {"jobs": [], "wct": sol["wct"]}
	# print("sol init : ", sol)

	for i in range(0, len(sol["jobs"])-1):
		temp_sol = copy.copy(sol["jobs"])
		swapValues(temp_sol, i, i+1)
		wct = instance.computeWCT3(temp_sol)
		if (wct < best_sol["wct"]):
			best_sol["jobs"] = copy.copy(temp_sol)
			best_sol["wct"] = wct
			if (pivoting_rule == "FI"): # stop directly after 1 sol found
				return best_sol

		# print(temp_sol, wct)

	return best_sol

def getBestExchangeNeighbour(sol, instance, pivoting_rule):
	"""
	Two permutations ø, ø' are 2-exchange neighbours if, and only if, 
	one can be obtained from the other by exchanging two jobs at arbitrary positions
	ex : [A,B,C,D,E,F]
	=>   [A,E,C,D,B,F]
	"""
	best_sol = {"jobs": [], "wct": sol["wct"]}
	# print("sol init : ", sol)

	for i in range(len(sol["jobs"])):
		for j in range(len(sol["jobs"])):
			if (i != j):
				temp_sol = copy.copy(sol["jobs"])
				swapValues(temp_sol, i, j)
				wct = instance.computeWCT3(temp_sol)
				if (wct < best_sol["wct"]):
					best_sol["jobs"] = copy.copy(temp_sol)
					best_sol["wct"] = wct
					if (pivoting_rule == "FI"): # stop directly after 1 sol found
						return best_sol

				# print(temp_sol, wct)
	return best_sol

def getBestInsertionNeighbour(sol, instance, pivoting_rule):
	"""
	Two permutations ø, ø' are insertion neighbours if, and only if, 
	one can be obtained from the other by removing a job from one position 
	and inserting it at another position
	ex : [A,B,C,D,E,F]
	=>   [A,C,D,B,E,F]

	QUESTION : l'insertion peuut se faire en fin de liste aussi ? 
	"""
	best_sol = {"jobs": [], "wct": sol["wct"]}
	# print("sol init : ", sol)

	for i in range(len(sol["jobs"])):
		for j in range(len(sol["jobs"])+1):
			if (i != j and j != i+1): #to avoid useless solutions
				temp_sol = copy.copy(sol["jobs"])
				temp_sol.insert(j, sol["jobs"][i])
				if (j < i):
					temp_sol.pop(i+1)
				else:
					temp_sol.pop(i)

				wct = instance.computeWCT3(temp_sol)
				if (wct < best_sol["wct"]):
					best_sol["jobs"] = copy.copy(temp_sol)
					best_sol["wct"] = wct
					if (pivoting_rule == "FI"): # stop directly after 1 sol found
						return best_sol

				# print(temp_sol, wct)
				
	return best_sol

def isLocalOptimal(sol):
	"""
	Check if the solution is a local optimum. By definition, a local optimum (here minimum) is :
	search position without improving neighbours w.r.t. given evaluation function g and neighbourhood
	N. i.e. position s in S such that g(s) <= g(s') for all s' of N(s). 
	=> same as checking if the neighbour found is empty. (if empty : sol is a local optimum)
	"""
	return sol["jobs"] == [] # if true => local optimum

def chooseNeighbour(sol, instance, neighbour_type, pivoting_rule):
	"""
	Choose one neighbour (the best one) among all the neighbours of a given solution. 
	"""
	if (neighbour_type == "T"):
		return getBestTransposeNeighbour(sol, instance, pivoting_rule) 
	elif (neighbour_type == "E"):
		return getBestExchangeNeighbour(sol, instance, pivoting_rule)
	elif (neighbour_type == "I"):
		return getBestInsertionNeighbour(sol, instance, pivoting_rule)

def iterativeImprovement(args, instance):
	"""
	Heuristic algorithm to find the optimal solution to the PSFP
	args = [name.py, initsol, pivotingrule, neighbourhood]
	"""
	solution = generateInitialSolution(args[1], instance) # dictionary
	neighbour = chooseNeighbour(solution, instance, args[3], args[2])
	it = 0
	while not isLocalOptimal(neighbour):
		solution["jobs"] = copy.copy(neighbour["jobs"])
		solution["wct"] = neighbour["wct"]
		# print("it: {0} | {1} : {2}".format(it, solution["jobs"], solution["wct"]))
		neighbour = chooseNeighbour(solution, instance, args[3], args[2])
		it += 1

	return solution

def variableNeighbourhoodDescent(initsolRule, neighbourhoods, instance):
	"""
	VND variable neighbourhood descent
	use different neighbourhood mode throughout the algorithm
	args = [name.py, initsol, pivotingrule, neighbourhood]
	"""

	solution = generateInitialSolution(initsolRule, instance) # dictionary
	i = 0
	it = 0
	while i > len(neighbourhoods):
		print("neighbourhood mode : ", n)
		neighbour = chooseNeighbour(solution, instance, neighbourhoods[i], "FI")
		if (isLocalOptimal(neighbour)): # if no existing improving solution
			i += 1
		else:
			solution["jobs"] = copy.copy(neighbour["jobs"])
			solution["wct"] = neighbour["wct"]
			print("it: {0} | {1} : {2}".format(it, solution["jobs"], solution["wct"]))
			i = 0
			it += 1

	return solution

def averageRelativePercentageDev(deltas):
	"""
	Returns the average relative percentage deviation of all instances of algorithm k
	algorithm k = a combination of different mode
		for example : --random --firstImprovement --tranpose
	"""
	return sum(delta) / len(delta) 


def readBestKnownWCT(filename):
	"""
	Read the instance best known WCT file
	filename = bestSolutions.txt
	"""

	best_knowns = {}
	try:
		with open(filename) as fileIn:
			lines = fileIn.readlines()[1:] # we ignore the first line 
			for line in lines:
				l = line.strip().split(" , ")
				best_knowns[l[0]] = float(l[1]) # filename : best score

	
	except FileNotFoundError:
		print("File does not exist : ", filename)
	except:
		print("Other error")

	return best_knowns

